resource/kind.go

```diff
@@ -0,0 +1,319 @@
+package resource
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var (
+	_ Kind    = &UntypedKind{}
+	_ Object2 = &UntypedObject{}
+)
+
+// Object2 implements kubernetes' runtime.Object and meta/v1.Object, as well as some additional methods useful for the app-sdk
+type Object2 interface {
+	runtime.Object
+	metav1.Object
+	GetSpec() any
+	GetSubresources() map[string]any
+	GetStaticMetadata() StaticMetadata
+	SetStaticMetadata(metadata StaticMetadata)
+}
+
+// MultiVersionKind is a collection of Kinds which have the same Group and Kind, but different versions.
+// This may allow us to include conversion as a component of the interface.
+// TODO: not sure if necessary
+type MultiVersionKind interface {
+	Group() string
+	Kind() string
+	Versions() []Kind
+}
+
+// Group is a collection of Kinds which share a Group and Version, allowing them to be put under the same
+// routing in an API server.
+// TODO: not sure if necessary
+type Group interface {
```

Comment:
It would be nice to have some examples of use case that make this useful

Reply:
Maybe the idea is something like (obviously this is not correct code):
```
groupV1 := &Group{ "slo.api.grafana.com", "v1", [SLOv1{}] }
groupV2 := &Group{ "slo.api.grafana.com", "v2", [SLOv2{}, UserPreferences{}]
router := &VersionedResourceRouter{}
router.HandleResourceGroup(groupV1)
router.HandleResourceGroup(groupV2)
```
and that sets up standard handlers for every type in a group. Don't know where the handler functions come from in that case, though.